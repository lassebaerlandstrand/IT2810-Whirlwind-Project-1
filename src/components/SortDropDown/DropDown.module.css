.container {
  width: 100%;
  position: relative;
}

.dropDownButton {
  width: 100%;
  padding: 0.5rem;
  background-color: white;
  border-radius: 0.3rem;
  border: 1px solid var(--inactive-border);
  cursor: pointer;
}
