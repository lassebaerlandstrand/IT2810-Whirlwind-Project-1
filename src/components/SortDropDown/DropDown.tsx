import React, { useEffect, useRef, useState } from 'react';
import styles from './DropDown.module.css';

type DropDownProps = {
  options: string[];
  selectedOption: string;
  setSelectedOption: (option: string) => void;
  label: string;
};

export const DropDown: React.FC<DropDownProps> = ({ selectedOption, options, setSelectedOption, label }) => {
  const [open, setOpen] = useState<boolean>(false);
  const dropDownRef = useRef<HTMLMenuElement>(null);

  const toggleOpen = () => setOpen((prev) => !prev);
  const handleOptionSelect = (option: string) => {
    setSelectedOption(option);
    setOpen(false);
  };

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (dropDownRef.current && !dropDownRef.current.contains(event.target as Node)) {
        setOpen(false);
      }
    };
    document.addEventListener('click', handleClickOutside);
    return () => document.removeEventListener('click', handleClickOutside);
  }, []);

  return (
    <div className={styles.container} ref={dropDownRef}>
      <label className={styles.label}>{label}</label>
      <button className={styles.dropDownButton} onClick={toggleOpen} aria-expanded={open} aria-haspopup="listbox">
        {selectedOption}
      </button>
      {open && (
        <ul className={styles.dropDownContent} role="listbox">
          {options.map((option) => (
            <li
              key={option}
              className={styles.dropDownItem}
              onClick={() => handleOptionSelect(option)}
              role="option"
              aria-selected={option === selectedOption}
            >
              {option}
            </li>
          ))}
        </ul>
      )}
    </div>
  );
};
