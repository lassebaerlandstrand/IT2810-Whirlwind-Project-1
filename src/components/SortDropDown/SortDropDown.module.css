.container {
  position: relative;
  margin-block-start: 0;
  margin-block-end: 0;
  padding-inline-start: 0;
}

.dropDownButton {
  display: flex;
  align-items: center;
  justify-content: space-between;
  width: fit-content;
  min-width: 11rem;
  padding: 0.5rem;
  background-color: white;
  border-radius: 0.3rem;
  border: 0.1rem var(--inactive-border) solid;
  cursor: pointer;
  box-sizing: border-box;
  user-select: none;
  font-size: 1rem;
}

.iconSpan {
  display: flex;
  align-items: center;
  justify-content: center;
  padding-left: 1rem;
  color: #999999;
}

.dropDownContent {
  display: none;
  flex-direction: column;
  align-items: center;
  position: absolute;
  min-width: 100%;
  padding: 0;
  background-color: white;
  max-height: 20vh;
  overflow-y: scroll;
  scrollbar-width: none;
  -ms-overflow-style: none;
  box-sizing: border-box;
  list-style-type: none;
  user-select: none;
  margin-top: 0.25rem;
  border: 0.1rem var(--inactive-border) solid;
  border-radius: 0.3rem;
}

.dropDownContent::-webkit-scrollbar {
  width: 0;
  height: 0;
}

.dropDownItem {
  width: 100%;
  padding: 0.5rem;
  cursor: pointer;
  box-sizing: border-box;
}

.dropDownItem:hover {
  background-color: #f4f4f4;
}

.buttonOpen {
  border: 0.1rem var(--active-border-blue) solid;
}

.contentOpen {
  display: flex;
}
