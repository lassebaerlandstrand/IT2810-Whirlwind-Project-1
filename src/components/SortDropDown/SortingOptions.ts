// Had to extract to a serparate file to fix eslint warning
import { Location } from '../../types/api-types';

type OptionsType = {
  [key: string]: (a: Location, b: Location) => number;
};

// TODO: Add more sorting options
export const options: OptionsType = {
  Alphabetically: (a: Location, b: Location) => a.city_name.localeCompare(b.city_name),
  Country: (a: Location, b: Location) => a.country_name.localeCompare(b.country_name),
};
