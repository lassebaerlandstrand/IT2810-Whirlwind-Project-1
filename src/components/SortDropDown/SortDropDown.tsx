import React, { useEffect, useState } from 'react';
import { options } from '../../contexts/LocationContext';
import DropDown from '../DropDown/DropDown';

type SortDropDownProps = {
  setSortKey: any;
};

const SortDropDown: React.FC<SortDropDownProps> = ({ setSortKey }) => {
  const [selectedOption, setSelectedOption] = useState<string>(
    () => sessionStorage.getItem('SortingOption') || Object.keys(options)[0],
  );

  useEffect(() => {
    sessionStorage.setItem('SortingOption', selectedOption);
    setSortKey(selectedOption);
  }, [selectedOption, setSortKey]);

  const optionList = Object.keys(options);

  return (
    <DropDown
      options={optionList}
      selectedOption={selectedOption}
      setSelectedOption={setSelectedOption}
      label="Sort by:"
    />
  );
};

export default SortDropDown;
