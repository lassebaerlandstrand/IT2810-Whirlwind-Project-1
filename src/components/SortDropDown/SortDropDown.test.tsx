import { fireEvent, render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import SortDropDown from './SortDropDown';
import { options } from './SortingOptions';

describe('SortDropDown Component', () => {
  const mockSetSortCondition = vi.fn();

  beforeEach(() => {
    sessionStorage.clear();
    vi.resetAllMocks();
  });

  it('should render with the default sorting option from session storage', () => {
    sessionStorage.setItem('SortingOption', 'Country Name');
    render(<SortDropDown setSortCondition={mockSetSortCondition} />);

    expect(screen.getByText('Country Name')).toBeInTheDocument();
    expect(mockSetSortCondition).toHaveBeenCalledWith(options['Country Name']);
  });

  it('should render with the default sorting option when session storage is empty', () => {
    render(<SortDropDown setSortCondition={mockSetSortCondition} />);

    expect(screen.getByText('City Name')).toBeInTheDocument();
    expect(mockSetSortCondition).toHaveBeenCalledWith(options['City Name']);
  });

  it('should save the selected option to session storage and update the sorting method', () => {
    render(<SortDropDown setSortCondition={mockSetSortCondition} />);

    fireEvent.click(screen.getByRole('button'));
    fireEvent.click(screen.getByText('Country Name'));

    expect(sessionStorage.getItem('SortingOption')).toBe('Country Name');
    expect(mockSetSortCondition).toHaveBeenCalledWith(options['Country Name']);
  });

  it('should open and close the dropdown menu when clicked', () => {
    render(<SortDropDown setSortCondition={mockSetSortCondition} />);

    const button = screen.getByRole('button', { name: 'City Name' });

    fireEvent.click(button);
    expect(screen.getByRole('listbox')).toBeInTheDocument();

    fireEvent.click(button);
    expect(screen.queryByRole('listbox')).not.toBeInTheDocument();
  });

  it('should match snapshot', () => {
    const { asFragment } = render(<SortDropDown setSortCondition={vi.fn()} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
