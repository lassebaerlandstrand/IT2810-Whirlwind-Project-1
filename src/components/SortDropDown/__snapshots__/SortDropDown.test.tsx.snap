// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SortDropDown Component > should match snapshot 1`] = `
<DocumentFragment>
  <menu
    class="_container_20b9ed"
  >
    <label
      class="_label_20b9ed"
    >
      Sort by:
    </label>
    <button
      aria-describedby="dropdown-button"
      aria-expanded="false"
      class="_dropDownButton_20b9ed "
      data-testid="dropdown-button"
    >
      City Name 
      <span
        class="_iconSpan_20b9ed"
      >
        <svg
          class="tabler-icon tabler-icon-chevron-down "
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6 9l6 6l6 -6"
          />
        </svg>
      </span>
    </button>
  </menu>
</DocumentFragment>
`;
